larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+/**
+ *  @file   larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc
+ *
+ *  @brief  Implementation of the hit width cluster merging algorithm class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArObjects/LArTwoDSlidingFitResult.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+
+HitWidthClusterMergingAlgorithm::HitWidthClusterMergingAlgorithm() :
+  m_clusterListName(),
+  m_maxConstituentHitWidth(0.5),
+  m_hitWidthScalingFactor(1.0),
+  m_useSlidingLinearFit(false),
+  m_layerFitHalfWindow(20),
+  m_fittingWeight(10),
+  m_minClusterWeight(0.5),      
+  m_maxXMergeDistance(5),     
+  m_maxZMergeDistance(2),     
+  m_maxMergeCosOpeningAngle(0.97),
+  m_maxDirectionDeviationCosAngle(0.90)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::GetListOfCleanClusters(const ClusterList *const pClusterList, ClusterVector &clusterVector) const
+{
+ 
+    LArHitWidthHelper::ClusterToParametersMapStore* pClusterToParametersMapStore = LArHitWidthHelper::ClusterToParametersMapStore::Instance();
+    LArHitWidthHelper::ClusterToParametersMap* pClusterToParametersMap = pClusterToParametersMapStore->GetMap();
+
+    // Clear map if already full i.e. from other view clustering
+    if(!pClusterToParametersMap->empty())
+        pClusterToParametersMap->clear();
+
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+        // The original cluster weight, with no hit scaling or hit padding
+        if(LArHitWidthHelper::GetOriginalTotalClusterWeight(pCluster) < m_minClusterWeight)
+            continue;
+
+        pClusterToParametersMap->insert(std::pair<const Cluster*, LArHitWidthHelper::ClusterParameters>(pCluster, LArHitWidthHelper::ClusterParameters(pCluster, m_maxConstituentHitWidth, m_useSlidingLinearFit, m_hitWidthScalingFactor)));
+
+        clusterVector.push_back(pCluster);
+    }
+
+    //ORDER BY MAX EXTREMAL X COORDINATE
```

Comment:
I'd argue this comment is superfluous (and may go out of date), and that the sort call below is self-describing, which is a nice thing to achieve.